app/controllers/concerns/manager_invitations.rb

```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module ManagerInvitations
+  extend ActiveSupport::Concern
+
+  def create_new_manager(email, enterprise)
+    password = Devise.friendly_token
+    new_user = Spree::User.create(email: email, unconfirmed_email: email, password: password)
+    new_user.reset_password_token = Devise.friendly_token
+    # Same time as used in Devise's lib/devise/models/recoverable.rb.
+    new_user.reset_password_sent_at = Time.now.utc
+    new_user.save
+
+    return new_user unless new_user.valid? # Return early if user is invalid.
+
+    enterprise.users << new_user
+    EnterpriseMailer.manager_invitation(@enterprise, new_user).deliver_later
```

Comment:
The `save` action is already calling `valid?`. Here we are running it twice which can be expensive. And sometime we have other checks in a `before_save` action which is not run by `valid?`. `Spree::User` doesn't seem to have such callbacks which means that we are okay here but the better way would be to check for errors occurred during save:

```rb
    return new_user if new_user.error.present?
```

But here you only look at returning early or avoiding adding to the enterprise. You can use the return value for that:

```suggestion
    if new_user.save
      enterprise.users << new_user
      EnterpriseMailer.manager_invitation(@enterprise, new_user).deliver_later
    end
```

Reply:
Thanks, that's super clear now.